"Deep dive into a Debezium community connector: Scylla CDC Source Connector" blogpost

avelanarius opened the pull request.

avelanarius commented:
cc: @gunnarmorling @PeterCorless

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```diff
@@ -0,0 +1,229 @@
+    0x41400... |  808e... | ... |             2 |        2 | blueberries |    Alice
+----
+
+All three operations are visible in the CDC log. There are two INSERTs (`cdc$operation = 2`) and one UPDATE (`cdc$operation = 1`). For each operation, its timestamp is also preserved (`cdc$time` column as time UUID).
```
How does a UUID represent a time? This might be worth an explanatory sentence, as it's not obvious (not to me, anyways).

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```diff
@@ -0,0 +1,229 @@
+author: pgrabowski
+---
+
+At ScyllaDB, we develop a high-performance NoSQL database https://www.scylladb.com/[Scylla], API-compatible with Apache Cassandra, Amazon DynamoDB and Redis. Earlier this year, we introduced support for https://docs.scylladb.com/using-scylla/cdc/cdc-intro/[Change Data Capture] in Scylla 4.3. This new feature seemed like a perfect match for integration with the Kafka ecosystem, so we developed the ...
```
```suggestion
At ScyllaDB, we develop a high-performance NoSQL database https://www.scylladb.com/[Scylla], API-compatible with Apache Cassandra, Amazon DynamoDB and Redis. Earlier this year, we introduced support for https://docs.scylladb.com/using-scylla/cdc/cdc-intro/[Change Data Capture] in Scylla 4.3. This new feature seemed like a perfect match for integration with the Apache Kafka ecosystem, so we developed the https://github.com/scylladb/scylla-cdc-source-connector[Scylla CDC Source Connector] using the Debezium framework. In this blogpost we will cover the basic structure of Scylla’s CDC, reasons we chose the Debezium framework and design decisions we made.
```

The first reference should be qualified, adhering to the standard usage rules/recommendations.

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
Nice, had no idea about that JavaDoc publication :) Let's only reference the latest one, 0.7.2 is really old.

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```suggestion
Thanks to this, when you start the Scylla CDC Source Connector, the configuration parameters will be immediately familiar, as many of them are common with other Debezium connectors. The generated data change events have the same https://javadoc.io/static/io.debezium/debezium-core/0.7.2/io/debezium/data/Envelope.html[Envelope] structure as generated by other Debezium connectors. This similarity allows for the use of many standard Debezium features, such as link:/documentation/reference/1.6/transformations/event-flattening.html[New Record State Extraction] [.underline]#.#
```

That way, it's a relative link, also working correctly with local website builds, etc.

gunnarmorling replied:
What is the underline part at the end about?

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
That's an interesting decision; how are those range deletes represented in the ScyllaDB CDC table? In case of other connectors, the Debezium connector would typically receive one delete event from the database for each deleted row (i.e. a row-based representation, instead of statement-based). Do you have plans to support deletes at some point in the future?

avelanarius replied:
Clarified this section in the fixup commit. Scylla performs range deletes by writing a tombstone. In the Scylla CDC log, two rows are inserted: one that indicates the start of the deletion range, one that indicates the end of the deletion range. Therefore it's not possible to read all rows that were deleted.

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```suggestion
However, most sink connectors won’t be able to correctly parse such a structure. Therefore, we decided to develop our own SMT, based on Debezium’s link:/documentation/reference/1.6/transformations/event-flattening.html[New Record State Extraction SMT]. Our https://github.com/scylladb/scylla-cdc-source-connector#scyllaextractnewstate-transformer[ScyllaExtractNewState] SMT works by applying Debezium's New Record State Extraction and flattening the `{"value": ...}` structures:
```

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
Isn't this exposing the same issue you seeked to circumvent with the nested representation? I.e. how will a consumer distinguish a field not modified from one whose value was set to null?

avelanarius replied:
Yes, this transformation is unfortunately lossy.

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```suggestion
Each created Kafka Connect task (that can run on a separate Kafka Connect worker node) reads CDC changes from its assigned set of streams. If you double the number of tasks, each task will have to read only a half of the number of streams - half of data throughput, making it possible to handle a higher load.
```

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
600 k/s of what? Transactions? Changed rows?

gunnarmorling commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
The reference to Confluent Platform comes a bit out of nowhere. How about: "With a setting of xyz (which is the default setting for instance on Confluent Platform), this topic ..."? Or just "With a setting of xyz, this topic...".

gunnarmorling reviewed (commented):
Thanks a lot, @avelanarius!

This looks really great! I've pushed one commit for fixing the reference to the image. Some small remarks/suggestions inline. Besides the question on deletes, I'm curious about snapshotting: is this supported by the ScyllaDB connector? If so, perhaps worth mentioning it, potentially also sharing details about how you ensure a consistent transition from the snapshot to stream reading phase?

In terms of publication: I'd suggest to publish this next Monday, as Friday is generally bad in terms of attention. I'd adjust the date of the post accordingly.

Thanks again, and looking forward to seeing this one live!

gunnarmorling commented:
Hey @avelanarius, any thoughts on the comments above?

avelanarius commented:
@gunnarmorling Thank you for your patience - unfortunately I was very busy on urgent issues last week. I have addressed all the feedback.

Regarding snapshotting: the connector does not support it at the moment. We are still at a phase that we figure out the approach we will take (to ensure the consistent transition and have good performance).

avelanarius commented on _posts/2021-09-08-deep-dive-into-a-debezium-community-connector-scylla-cdc-source-connector.adoc:
```diff
@@ -0,0 +1,229 @@
+Alice |        2 | blueberries
+----
+
+Looking only at this table, you cannot easily reconstruct all modifications that took place, e.g. Tim’s order name before the UPDATE. Let’s look at the CDC log, easily accessible as a table (some columns truncated for clarity):
```
That's correct - cannot at all.

gunnarmorling commented:
Rebased and applied, should be live in a few min. Thanks a lot, @avelanarius!
